### PR TITLE
dialects: Add LTL dialect (AndOp, Sequence and Property types)

### PR DESCRIPTION
**Pull request by luisacicolini**
Added And Operator from LTL dialect (https://circt.llvm.org/docs/Dialects/LTL/#ltland-circtltlandop) + some types and relative testing
I'm not sure the verification is necessary, there might be a more concise way to specify that operand and result types must be the same, but I am not sure how to do that

### CODE REVIEW

File: xdsl/dialects/ltl.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+            self, 
+            operand: SSAValue, 
+    ):
+        super().__init__(operands = [operand])
+
+    def verify_(self):
+        if isinstance(self.input, property) ^ isinstance(self.result, property):
+            raise ValueError("AndOp: property type mismatch")
+        if isinstance(self.input, sequence) ^ isinstance(self.result, sequence):
+            raise ValueError("AndOp: sequence type mismatch")
```
Comment (superlopuh) [inline xdsl/dialects/ltl.py:88]:
```suggestion
```
This is already verified by the constraint variable.

File: xdsl/dialects/ltl.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    value (`i1`) or a sequence (`ltl.sequence`) is also a valid property.
+    Operations that accept a property as an operand will use the `AnyProperty`
+    constraint, which also accepts `ltl.sequence` and `i1`.
+    """
+
+    name = "ltl.propertytype"
+
+
+@irdl_attr_definition
+class sequence(ParametrizedAttribute, TypeAttribute):
```
Comment (superlopuh) [inline xdsl/dialects/ltl.py:49]:
```suggestion
class Sequence(ParametrizedAttribute, TypeAttribute):
```
Please use PascalCase for class names

File: xdsl/dialects/ltl.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    """
+    The ltl.sequence type represents a sequence of linear temporal logic, for example,
+    “A is true two cycles after B is true”. Note that this type explicitly identifies
+    a sequence. However, a boolean value (i1) is also a valid sequence.
+    Operations that accept a sequence as an operand will use the AnySequence constraint,
+    which also accepts i1.
+    """
+
+    name = "ltl.sequencetype"
```
Comment (superlopuh) [inline xdsl/dialects/ltl.py:58]:
```suggestion
    name = "ltl.sequence"
```

File: xdsl/dialects/ltl.py [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+
+
+"""
+Implementation of the LTL dialect by CIRCT. Documentation: https://circt.llvm.org/docs/Dialects/LTL/
+"""
+
+
+@irdl_attr_definition
+class property(ParametrizedAttribute, TypeAttribute):
```
Comment (superlopuh) [inline xdsl/dialects/ltl.py:34]:
```suggestion
class Property(ParametrizedAttribute, TypeAttribute):
```

File: xdsl/dialects/ltl.py [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+    The `ltl.property` type represents a verifiable property built from linear
+    temporal logic sequences and quantifiers, for example, *"if you see sequence
+    A, eventually you will see sequence B"*.
+    Note that this type explicitly identifies a *property*. However, a boolean
+    value (`i1`) or a sequence (`ltl.sequence`) is also a valid property.
+    Operations that accept a property as an operand will use the `AnyProperty`
+    constraint, which also accepts `ltl.sequence` and `i1`.
+    """
+
+    name = "ltl.propertytype"
```
Comment (superlopuh) [inline xdsl/dialects/ltl.py:45]:
```suggestion
    name = "ltl.property"
```

File: xdsl/dialects/ltl.py [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+    “A is true two cycles after B is true”. Note that this type explicitly identifies
+    a sequence. However, a boolean value (i1) is also a valid sequence.
+    Operations that accept a sequence as an operand will use the AnySequence constraint,
+    which also accepts i1.
+    """
+
+    name = "ltl.sequencetype"
+
+
+ltltype = ContainerOf(AnyOf([sequence, property, signlessIntegerLike]))
```
Comment (superlopuh) [inline xdsl/dialects/ltl.py:61]:
```suggestion
LTL_TYPE = ContainerOf(AnyOf([sequence, property, signlessIntegerLike]))
```
Please use UPPER_SNAKE_CASE for top-level constants

File: tests/filecheck/dialects/ltl/ltl_basic.mlir [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+// CHECK: unrealized_conversion_cast to !ltl.property
+%s = unrealized_conversion_cast to !ltl.sequence
+%p = unrealized_conversion_cast to !ltl.property
+
+// CHECK-NEXT: ltl.and {{%.+}}, {{%.+}} : i1, i1
+// CHECK-NEXT: ltl.and {{%.+}}, {{%.+}} : !ltl.sequence, !ltl.sequence
+// CHECK-NEXT: ltl.and {{%.+}}, {{%.+}} : !ltl.property, !ltl.property
+ltl.and %true, %true : i1, i1
+ltl.and %s, %s : !ltl.sequence, !ltl.sequence
+ltl.and %p, %p : !ltl.property, !ltl.property
```
Comment (superlopuh) [inline tests/filecheck/dialects/ltl/ltl_basic.mlir:15]:
```suggestion
ltl.and %p, %p : !ltl.property, !ltl.property

```
Please add newlines at the end of files

**Review by superlopuh — CHANGES_REQUESTED** [review]

**Comment by superlopuh** [comment]
Ah sorry noticed this was a draft after I reviewed, hope the comments are useful anyway, even if this is a work in progress.

**Comment by luisacicolini** [comment]
> Ah sorry noticed this was a draft after I reviewed, hope the comments are useful anyway, even if this is a work in progress.

They sure are! Thanks a lot for your time :) I wanted to submit it but then noticed some major issues with a test so drafted it back!

File: tests/filecheck/dialects/ltl/ltl_invalid.mlir [hunk 8]
```diff
@@ -0,0 +1,13 @@
+// RUN XDSL_ROUNDTRIP
+
+// CHECK: attribute !ltl.sequence expected from variable 'T', but got !ltl.property
+
+"builtin.module"() ({
+
```
Comment (compor) [inline tests/filecheck/dialects/ltl/ltl_invalid.mlir:6]: Could you please remove extraneous whitespace/empty lines?

File: tests/filecheck/dialects/ltl/ltl_invalid.mlir [hunk 9]
```diff
@@ -0,0 +1,13 @@
+// RUN XDSL_ROUNDTRIP
+
+// CHECK: attribute !ltl.sequence expected from variable 'T', but got !ltl.property
```
Comment (compor) [inline tests/filecheck/dialects/ltl/ltl_invalid.mlir:3]: Could you move this after the `module` or even better right after the statement that is invalid and raises this exception?

File: tests/filecheck/dialects/ltl/ltl_op.mlir [hunk 10]
```diff
@@ -0,0 +1,19 @@
+// RUN XDSL_ROUNDTRIP
+
+// CHECK:       "builtin.module"() ({
```
Comment (compor) [inline tests/filecheck/dialects/ltl/ltl_op.mlir:3]: Same here, please move after the `module`.

File: xdsl/dialects/ltl.py [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+
+@irdl_attr_definition
+class Property(ParametrizedAttribute, TypeAttribute):
+    """
+    The `ltl.property` type represents a verifiable property built from linear
+    temporal logic sequences and quantifiers, for example, *"if you see sequence
+    A, eventually you will see sequence B"*.
+    Note that this type explicitly identifies a *property*. However, a boolean
+    value (`i1`) or a sequence (`ltl.sequence`) is also a valid property.
+    Operations that accept a property as an operand will use the `AnyProperty`
```
Comment (AntonLydike) [inline xdsl/dialects/ltl.py:30]: I guess `AnyProperty` is no longer a thing? I would keep implementation details out of docstrings, they are usually not as helpful as describing what happens in more concise terms.

File: xdsl/dialects/ltl.py [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+    name = "ltl.property"
+
+
+@irdl_attr_definition
+class Sequence(ParametrizedAttribute, TypeAttribute):
+    """
+    The ltl.sequence type represents a sequence of linear temporal logic, for example,
+    “A is true two cycles after B is true”. Note that this type explicitly identifies
+    a sequence. However, a boolean value (i1) is also a valid sequence.
+    Operations that accept a sequence as an operand will use the AnySequence constraint,
```
Comment (AntonLydike) [inline xdsl/dialects/ltl.py:43]: Same here!

File: xdsl/dialects/ltl.py [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+    IRDLOperation,
+    irdl_attr_definition,
+    irdl_op_definition,
+    result_def,
+    var_operand_def,
+)
+
+"""
+Implementation of the LTL dialect by CIRCT. Documentation: https://circt.llvm.org/docs/Dialects/LTL/
+"""
```
Comment (compor) [inline xdsl/dialects/ltl.py:19]:
```suggestion
"""
Implementation of the LTL dialect by CIRCT.

https://circt.llvm.org/docs/Dialects/LTL/
"""
```

Comment (compor) [reply xdsl/dialects/ltl.py:19]: Moreover, module docstrings should be at the very top of the file.
Please move it there.

File: xdsl/dialects/ltl.py [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+    """
+
+    name = "ltl.sequence"
+
+
+@irdl_op_definition
+class AndOp(IRDLOperation):
+    """
+    A conjunction of booleans, sequences, or properties. If any of the $inputs is of type
+    !ltl.property, the result of the op is an !ltl.property. Otherwise it is an !ltl.sequence.
```
Comment (AntonLydike) [inline xdsl/dialects/ltl.py:54]: I feel you could work on all three of your docstrings, This reads as if they are either sequence or property, but can't they also be `i1`?

File: xdsl/dialects/ltl.py [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+        operand: SSAValue,
+    ):
+        super().__init__(operands=[operand])
+
+
+LTL = Dialect(
+    [
+        AndOp,
+    ],
+    [Property, Sequence],
```
Comment (AntonLydike) [inline xdsl/dialects/ltl.py:80]: supernit: 
```suggestion
    [
        Property, 
        Sequence,
    ],
```

**Review by AntonLydike — APPROVED** [review]
Looks good overall, just some nitpicks with your docstrings!

File: xdsl/dialects/ltl.py [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+    """
+    A conjunction of booleans, sequences, or properties. If any of the $inputs is of type
+    !ltl.property, the result of the op is an !ltl.property. Otherwise it is an !ltl.sequence.
+    """
+
+    name = "ltl.and"
+
+    T = Annotated[
+        Attribute,
+        AnyOf([Sequence, Property, IntegerType(1, signedness=Signedness.SIGNLESS)]),
```
Comment (compor) [inline xdsl/dialects/ltl.py:61]:
```suggestion
        AnyOf([Sequence, Property, IntegerType(1)]),
```

The default signedness is signless.

File: xdsl/dialects/ltl.py [hunk 17]
```diff
@@ -0,0 +1,81 @@
...
+"""
+
+
+@irdl_attr_definition
+class Property(ParametrizedAttribute, TypeAttribute):
+    """
+    The `ltl.property` type represents a verifiable property built from linear
+    temporal logic sequences and quantifiers, for example, *"if you see sequence
+    A, eventually you will see sequence B"*.
+    Note that this type explicitly identifies a *property*. However, a boolean
```
Comment (compor) [inline xdsl/dialects/ltl.py:28]:
```suggestion
    temporal logic sequences and quantifiers, for example,
    *"if you see sequence A, eventually you will see sequence B"*.
    Note that this type explicitly identifies a *property*. However, a boolean
```

File: tests/filecheck/dialects/ltl/ltl_invalid.mlir (hunk 9 above)
Comment (compor) [reply tests/filecheck/dialects/ltl/ltl_invalid.mlir:3]: Where is exception thrown from? I can't seem to find it in the dialect file.
Moreover, "variable 'T'" sounds like a Python-y generic message, which one shouldn't really care that much.

Could you please provide a human-legible error message that denotes which parts of the operation types should be in agreement?

**Review by compor — CHANGES_REQUESTED** [review]

File: tests/filecheck/dialects/ltl/ltl_invalid.mlir (hunk 9 above)
Comment (luisacicolini) [reply tests/filecheck/dialects/ltl/ltl_invalid.mlir:3]: The message is given by the Annotated, should I use a custom verify_ instead?

Comment (compor) [reply tests/filecheck/dialects/ltl/ltl_invalid.mlir:3]: Hmm, let me have another look.

Comment (compor) [reply tests/filecheck/dialects/ltl/ltl_invalid.mlir]: I think it's given by the `ConstraintVar` via PyRDL.
Anyway, that is fine then.

File: tests/filecheck/dialects/ltl/ltl_invalid.mlir [hunk 18]
```diff
@@ -0,0 +1,13 @@
...
+
+"builtin.module"() ({
+
+    %true = "arith.constant"() {value = true} : () -> i1
```
Comment (compor) [inline tests/filecheck/dialects/ltl/ltl_invalid.mlir:5]: Is `%true` used? if not, please remove.

File: tests/filecheck/dialects/ltl/ltl_op.mlir [hunk 19]
```diff
@@ -0,0 +1,19 @@
+// RUN XDSL_ROUNDTRIP
+
+"builtin.module"() ({
+    %b = "arith.constant"() {value = 0} : () -> i1
+    %seq = "arith.constant"() {value = 0} : () -> !ltl.sequence
+    %p = "arith.constant"() {value = 0} : () -> !ltl.property
+    "ltl.and"(%seq, %seq) : (!ltl.sequence,!ltl.sequence)->(!ltl.sequence)
+    "ltl.and"(%p, %p) : (!ltl.property,!ltl.property)->(!ltl.property)
+    "ltl.and"(%b, %b) : (i1,i1)->(i1)
+}) : () -> ()
```
Comment (superlopuh) [inline tests/filecheck/dialects/ltl/ltl_op.mlir:10]:
```suggestion
builtin.module {
    %b, %seq, %p = "test.op"() : () -> (i1, !ltl.sequence, !ltl.property)
    "ltl.and"(%seq, %seq) : (!ltl.sequence,!ltl.sequence)->(!ltl.sequence)
    "ltl.and"(%p, %p) : (!ltl.property,!ltl.property)->(!ltl.property)
    "ltl.and"(%b, %b) : (i1,i1)->(i1)
}) : () -> ()
```
We try to use test ops for operations where the source of the SSA value is not directly relevant to the test. In this case, arith is not being tested, so we wouldn't want this test to fail if we break arith.

Comment (superlopuh) [reply tests/filecheck/dialects/ltl/ltl_op.mlir:10]: Similarly in other test

Comment (luisacicolini) [reply tests/filecheck/dialects/ltl/ltl_op.mlir:10]: oh right! makes absolutely sense, correcting it rn

File: tests/filecheck/dialects/ltl/ltl_invalid.mlir [hunk 20]
```diff
@@ -0,0 +1,9 @@
+// RUN XDSL_ROUNDTRIP
+
+
+"builtin.module"() ({
+    %seq = "arith.constant"() {value = 0} : () -> !ltl.sequence
```
Comment (superlopuh) [inline tests/filecheck/dialects/ltl/ltl_invalid.mlir:5]: same here please :)

File: xdsl/dialects/ltl.py [hunk 21]
```diff
@@ -0,0 +1,79 @@
...
+    “A is true two cycles after B is true”.
+    """
+
+    name = "ltl.sequence"
+
+
+@irdl_op_definition
+class AndOp(IRDLOperation):
+    """
+    A conjunction of booleans, sequences, or properties. If any of the $inputs is of type
```
Comment (superlopuh) [inline xdsl/dialects/ltl.py:48]: nit:
```suggestion
    A conjunction of booleans, sequences, or properties. If any of the `inputs` is of type
```

**Review by superlopuh — APPROVED** [review]
Thank you! Can you update the second test, please?

**Review by superlopuh — APPROVED** [review]

**Review by webmiche — APPROVED** [review]

File: tests/filecheck/dialects/ltl/ltl_invalid.mlir [hunk 22]
```diff
@@ -0,0 +1,8 @@
+// RUN XDSL_ROUNDTRIP
+
+
```
Comment (compor) [inline tests/filecheck/dialects/ltl/ltl_invalid.mlir:3]:
```suggestion

```

**Review by compor — APPROVED** [review]

**Pull request merged**